### DISCUSSION TITLE
Flash Sales on Custom Stores

### DISCUSSION THREAD

**Discussion by julianbenegas** [Shopify/storefront-api-feedback#33] [Help] [answered]
Hi, after reading [the documentation](https://shopify.dev/api/usage/rate-limits#storefront-api-rate-limits), it's not clear to me if the Storefront API is able to handle high loads of traffic.

> The bucket size and leak rate properties determine the API’s burst behavior and request rate. The 60-second limit applies to the IP address of the buyer interacting with the app, rather than the ID of the store where the app is installed. This means your app won't be throttled due to an increase in buyer traffic on a merchant’s store.

Then, in [the FAQs](https://github.com/Shopify/storefront-api-feedback/discussions/5), it says:

> Can I do a flash sale using the Storefront API?
> The Storefront API is not designed to be used for flash sales for two main reasons:
> 1. Bot protection is unavailable
> 2. The Storefront API has a limit on the amount of checkouts that can be created per minute. If an API client exceeds this throttle, then a 200 Throttled error response is returned. Shopify recommends designing your app to be resilient to this scenario. For example, you could implement a request queue with an exponential backoff algorithm.

Regarding #2 (`The Storefront API has a limit on the amount of checkouts that can be created per minute`): Is this limit known? **Does this mean that Shopify can't handle Flash Sales for custom stores, no matter how you design your app?** Or is there a way (with an Enterprise plan or something) to give more resources to the Storefront API?

Thanks

**Comment by vixdug** [comment, marked as answer]
Hey @julianbenegas - There are a few considerations when running a large flash sale on a custom storefront / Storefront API-powered app.  

Rate Limitis: 
1.  IP: As you've noted, the Storefront API is rate limited per buyer IP and therefore designed to be used client-side where the unique IP of the customer acts as a single user. This helps the SFAPI scale to support large amounts of requests / traffic. If you use this API through a proxy or server-side during a flash sale, all requests will come from a single IP and you will hit the throttle. 
2. Checkout: There is a limitation on the amount of CPM you can process both on the online store and the SFAPI. A typical large flash sale on the online store you would have the queue page get triggered and users are queued in a line to checkout. What the documentation is alluding to is a lack of queue if using the checkout mutations. This however is not the suggested way to create a checkout and is outdated. 

What you should be using is the **cart** flow which we announced at Unite. Here is a link: https://shopify.dev/custom-storefronts/cart during a flash sale the cart flow will redirect users to a checkout queue if necessary under high amount of load and is not subject to the hard checkout throttle mentioned above. 

TLDR: yes you should be able to do a flash sale on a custom storefront if you've developed the app correctly and are using the cart.

**Reply by julianbenegas** [reply]
Woah, that's amazing!

> during a flash sale the cart flow will redirect users to a checkout queue if necessary under high amount of load and is not subject to the hard checkout throttle mentioned above.

Let me know if I understood correctly: with high traffic, users will be able to add things to cart and such BUT will get into a queue once they go to the `checkoutUrl`.

We were thinking about creating a custom queue to handle SFAPI checkout limits, but it seems that this is solved with what you are saying 😮

**Reply by morganmccunn** [reply]
Yes that's exactly correct! No custom queue should be required :)

**Comment by morganmccunn** [comment]
We will update the FAQ, thanks for the flag :)
